### PR TITLE
Add a tutorial for working with table inputs in PyGMT

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Explain the table input types as mentioned in https://github.com/GenericMappingTools/pygmt/issues/1268.

**Preview**: https://pygmt-dev--2722.org.readthedocs.build/en/2722/get_started/04_table_inputs.html
**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: examples/get_started/04_working_with_tables.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+fig.plot(data="input_data.dat", style="p0.2c", fill="blue")
+fig.show()
+
+from pathlib import Path
+
+Path("input_data.dat").unlink()
+
+# %%
+# Besides a plain string to a table file, following variants are also accepted:
```
Comment (seisman) [inline examples/get_started/04_working_with_tables.py:49]: Need changes after https://github.com/GenericMappingTools/pygmt/issues/2723.

File: examples/get_started/04_working_with_tables.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
+"""
+4. Working with table inputs
+============================
+
+Generally, PyGMT accepts two different types of data inputs: tables and grids.
+
+- A table is a 2D array of data, with *M* rows and *N* columns. Each column
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:7]:
```suggestion
- A table is a 2-D array of data, with *M* rows and *N* columns. Each column 
```

File: examples/get_started/04_working_with_tables.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+"""
+4. Working with table inputs
+============================
+
+Generally, PyGMT accepts two different types of data inputs: tables and grids.
+
+- A table is a 2D array of data, with *M* rows and *N* columns. Each column 
+  represents a different variable (e.g., *x*, *y* and *z*) and each row 
+  represents a different record. 
+- A grid is a 2D array of data that is regularly spaced in the x and y directions.
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:10]:
```suggestion
- A grid is a 2-D array of data that is regularly spaced in the x and y directions. 
```

File: examples/get_started/04_working_with_tables.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+table input types available in PyGMT and provide examples for each. 
+By understanding the different table input types, you can choose the one that best fits 
+your data and analysis needs, and work more efficiently with PyGMT.
+"""
+
+# %%
+# ASCII table file
+# ----------------
+#
+# Most PyGMT functions/methods that accept table input data has a ``data`` parameter.
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:26]:
```suggestion
# Most PyGMT functions/methods that accept table input data have a ``data`` parameter.
```

File: examples/get_started/04_working_with_tables.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+# The easiest way to provide table input data to PyGMT is by specifying the
+# file name of an ASCII table (e.g., `data="table-input.txt"`). This is useful
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:28]: Maybe use the same file name as in the code example below.
```suggestion
# file name of an ASCII table (e.g., ``data="input_data.dat"``). This is useful
```

File: examples/get_started/04_working_with_tables.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+Path("input_data.dat").unlink()
+
+# %%
+# Besides a plain string to a table file, following variants are also accepted:
+#
+# - a `pathlib.Path` object
+# - a full URL. PyGMT will download the file to the current directory first.
+# - a file name prefixed with `@` (e.g., `data="@input_data.dat"`), which is
+#   a special syntax in GMT to indicate that the file is a remote file
+#   hosted on the GMT data server
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:55]:
```suggestion
# - A `pathlib.Path` object.
# - A full URL. PyGMT will download the file to the current directory first.
# - A file name prefixed with ``@`` (e.g., ``data="@input_data.dat"``), which is
#   a special syntax in GMT to indicate that the file is a remote file
#   hosted on the GMT data server.
```

File: examples/get_started/04_working_with_tables.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+# %%
+# 2-D Numpy array or pandas.DataFrame
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:58]:
```suggestion
# 2-D numpy array or pandas.DataFrame
```

File: examples/get_started/04_working_with_tables.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+# -----------------------------------
+#
+# The `data` parameter also accepts a 2D numpy array or a pandas.DataFrame object.
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:61]:
```suggestion
# The ``data`` parameter also accepts a 2-D numpy array or a pandas.DataFrame object.
```

File: examples/get_started/04_working_with_tables.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+# This is useful when you want to plot data that is already in memory.
+
+import pandas as pd
+
+fig = pygmt.Figure()
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+
+# Pass a 2D numpy array to the `data` parameter
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:69]:
```suggestion
# Pass a 2-D numpy array to the data parameter
```

File: examples/get_started/04_working_with_tables.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+fig.plot(data=np.array([[1.0, 2.0], [5.0, 4.0]]), style="t0.2c", fill="red")
+
+# Pass a pandas.DataFrame to the `data` parameter
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:72]:
```suggestion
# Pass a pandas.DataFrame to the data parameter
```

File: examples/get_started/04_working_with_tables.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+fig.plot(data=df, style="a0.5c", fill="blue")
+
+fig.show()
+
+# %%
+#
+# Scalar values or 1-D arrays
+# ---------------------------
+#
+# In addition to the `data` parameter, some PyGMT functions/methods also provide
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:83]:
```suggestion
# In addition to the ``data`` parameter, some PyGMT functions/methods also provide
```

File: examples/get_started/04_working_with_tables.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+# invididual parameters (e.g., `x` and `y` for data coordinates) which allow you
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:84]:
```suggestion
# invididual parameters (e.g., ``x`` and ``y`` for data coordinates) which allow you
```

File: examples/get_started/04_working_with_tables.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+# to specify the data. These parameters accept individual scalar values or 1-D arrays
+# (lists or 1-D numpy arrays). This is useful if you want to plot a single data
+# point or already have arrays of data in memory.
+#
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:88]:
```suggestion
```

File: examples/get_started/04_working_with_tables.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+fig.plot(data=np.array([[1.0, 2.0], [5.0, 4.0]]), style="t0.2c", fill="red")
+
+# Pass a pandas.DataFrame to the `data` parameter
+df = pd.DataFrame(np.array([[1.0, 3.0], [5.0, 2.0]]), columns=["x", "y"])
+fig.plot(data=df, style="a0.5c", fill="blue")
+
+fig.show()
+
+# %%
+#
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:79]:
```suggestion

```

File: examples/get_started/04_working_with_tables.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+# when your data is stored in a separate text file.
+
+import numpy as np
+import pygmt
+
+# Create an example file with 3 rows and 2 columns
+data = np.array([[1.0, 2.0], [5.0, 4.0], [8.0, 3.0]])
+np.savetxt("input_data.dat", data, fmt="%f")
+
+# Pass the file name to the `data` parameter
```
Comment (yvonnefroehlich) [inline examples/get_started/04_working_with_tables.py:38]:
```suggestion
# Pass the file name to the data parameter
```

**Review by yvonnefroehlich — COMMENTED** [review]
Just made some typo and formating comments. Have to look at the conent in more detail later.

File: examples/get_started/04_table_inputs.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+#   is a special syntax in GMT to indicate that the file is a remote file
+#   hosted on the GMT data server.
+
+# %%
+# 2-D array: list, numpy.ndarray, and pandas.DataFrame
+# ----------------------------------------------------
+#
+# The ``data`` parameter also accepts a 2-D array, e.g.,
+#
+# - A list of list
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:65]:
```suggestion
# - A list of two lists
```
or
```suggestion
# - A 2-D list
```

File: examples/get_started/04_table_inputs.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+# Example GeoDataFrame
+gdf = gpd.GeoDataFrame(
+    {
+        "geometry": gpd.points_from_xy([1, 2, 3], [2, 3, 4]),
+        "value": [10, 20, 30],
+    }
+)
+
+# Use the GeoDataFrame to specify the data
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:107]:
```suggestion
# Use the GeoDataFrame to specify the 'data' parameter
```

File: examples/get_started/04_table_inputs.py [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+np.savetxt("input_data.dat", data, fmt="%f")
+
+# Pass the file name to the data parameter
+fig = pygmt.Figure()
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+fig.plot(data="input_data.dat", style="p0.2c", fill="blue")
+fig.show()
+
+# Now let's delete the example file
+from pathlib import Path
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:46]: It seems like all import have to happen a the top of the file:

```
examples/get_started/04_table_inputs.py:46:1: E402 Module level import not at top of file
   |
45 | # Now let's delete the example file
46 | from pathlib import Path
   | ^^^^^^^^^^^^^^^^^^^^^^^^ E402
47 | 
48 | Path("input_data.dat").unlink()
   |

examples/get_started/04_table_inputs.py:71:1: E402 Module level import not at top of file
make: *** [Makefile:72: check] Error 1
   |
69 | # This is useful when you want to plot data that is already in memory.
70 | 
71 | import pandas as pd
   | ^^^^^^^^^^^^^^^^^^^ E402
72 | 
73 | fig = pygmt.Figure()
   |

examples/get_started/04_table_inputs.py:97:1: E402 Module level import not at top of file
   |
95 | # (e.g., GeoJSON, etc.) that GMT and PyGMT do not support natively.
96 | 
97 | import geopandas as gpd
   | ^^^^^^^^^^^^^^^^^^^^^^^ E402
98 | 
99 | # Example GeoDataFrame
   |
```

File: examples/get_started/04_table_inputs.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+# ASCII table file
+# ----------------
+#
+# Most PyGMT functions/methods that accept table input data have a ``data``
+# parameter. The easiest way to provide table input data to PyGMT is by
+# specifying the file name of an ASCII table (e.g., ``data="input_data.dat"``).
+# This is useful when your data is stored in a separate text file.
+
+import numpy as np
+import pygmt
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:33]: Hope this is the correct order...

```suggestion
from pathlib import Path

import geopandas as gpd
import numpy as np
import pandas as pd
import pygmt
```

File: examples/get_started/04_table_inputs.py (hunk 18 above)
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:46]:
```suggestion
```

File: examples/get_started/04_table_inputs.py [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+
+# Pass the file name to the data parameter
+fig = pygmt.Figure()
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+fig.plot(data="input_data.dat", style="p0.2c", fill="blue")
+fig.show()
+
+# Now let's delete the example file
+from pathlib import Path
+
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:47]:
```suggestion
```

File: examples/get_started/04_table_inputs.py [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+# ----------------------
+#
+# If you're working with geospatial data, you can read your data as a
+# :class:`geopandas.GeoDataFrames` object and pass it to the ``data``
+# parameter. This is useful if your data is stored in a geospatial data format
+# (e.g., GeoJSON, etc.) that GMT and PyGMT do not support natively.
+
+import geopandas as gpd
+
+# Example GeoDataFrame
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:99]:
```suggestion
# parameter. This is useful if your data is stored in a geospatial data format
# (e.g., GeoJSON, etc.) that GMT and PyGMT do not support natively.

# Example GeoDataFrame
```

File: examples/get_started/04_table_inputs.py [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+#
+# - A list of list
+# - A :class:`numpy.ndarray` object with with a dimension of 2
+# - A :class:`pandas.DataFrame` object
+#
+# This is useful when you want to plot data that is already in memory.
+
+import pandas as pd
+
+fig = pygmt.Figure()
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:73]:
```suggestion
# This is useful when you want to plot data that is already in memory.

fig = pygmt.Figure()
```

File: examples/get_started/04_table_inputs.py [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+4. PyGMT I/O: Table inputs
+==========================
+
+Generally, PyGMT accepts two different types of data inputs: tables and grids.
+
+- A table is a 2-D array of data, with *M* rows and *N* columns. Each column
+  represents a different variable (e.g., *x*, *y* and *z*) and each row
+  represents a different record.
+- A grid is a 2-D array of data that is regularly spaced in the x and y
+  directions.
```
Comment (seisman) [inline examples/get_started/04_table_inputs.py:11]: Need to have more accurate and precise definitions for tables and grids.

File: examples/get_started/04_table_inputs.py [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+fig.plot(data=gdf, style="c0.2c", fill="purple")
+fig.show()
+
+# %%
+# Scalar values or 1-D arrays
+# ---------------------------
+#
+# In addition to the ``data`` parameter, some PyGMT functions/methods also
+# provide invididual parameters (e.g., ``x`` and ``y`` for data coordinates)
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:116]:
```suggestion
# provide individual parameters (e.g., ``x`` and ``y`` for data coordinates)
```

File: examples/get_started/04_table_inputs.py [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+in the following tutorials.
+
+PyGMT supports a variety of table input types that allow you to work with data
+in a format that suits your needs. In this tutorial, we'll explore the
+different table input types available in PyGMT and provide examples for each.
+By understanding the different table input types, you can choose the one that
+best fits your data and analysis needs, and work more efficiently with PyGMT.
+"""
+
+#%%
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:23]:
```suggestion
# %%
```

File: examples/get_started/04_table_inputs.py [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+Generally, PyGMT accepts two different types of data inputs: tables and grids.
+
+- A table is a 2-D array with rows and columns. Each column represents a 
+  different variable (e.g., *x*, *y* and *z*) and each row represents a 
+  different record.
+- A grid is a 2-D array of data that is regularly spaced in the x and y
+  directions (or longitude and latitude). 
+
+In this tutorial, we'll focus on working with table inputs, and cover grid inputs
+in the following tutorials.
```
Comment (yvonnefroehlich) [inline examples/get_started/04_table_inputs.py:14]: Or do we plan to have multiple tutorials for grid inputs?
```suggestion
In this tutorial, we'll focus on working with table inputs, and cover grid
inputs in a separate tutorial.
```

File: examples/get_started/04_table_inputs.py [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+fig = pygmt.Figure()
+fig.basemap(region=[0, 10, 0, 5], projection="X10c/5c", frame=True)
+fig.plot(data="input_data.dat", style="p0.2c", fill="blue")
+fig.show()
+
+# Now let's delete the example file
+Path("input_data.dat").unlink()
+
+# %%
+# Besides a plain string to a table file, following variants are also accepted:
```
Comment (michaelgrund) [inline examples/get_started/04_table_inputs.py:54]:
```suggestion
# Besides a plain string to a table file, the following variants are also accepted:
```

File: examples/get_started/04_table_inputs.py [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+fig.show()
+
+# %%
+# Conclusion
+# ----------
+#
+# In PyGMT, you have the flexibility to provide data in various table input
+# types, including file names, 2-D array (2-D list, :class:`numpy.ndarray`,
```
Comment (michaelgrund) [inline examples/get_started/04_table_inputs.py:142]:
```suggestion
# types, including file names, 2-D arrays (2-D list, :class:`numpy.ndarray`,
```

File: examples/get_started/04_table_inputs.py [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+fig.show()
+
+# %%
+# :class:`geopandas.GeoDataFrame`
+# -------------------------------
+#
+# If you're working with geospatial data, you can read your data as a
+# :class:`geopandas.GeoDataFrame` object and pass it to the ``data``
+# parameter. This is useful if your data is stored in a geospatial data format
+# (e.g., GeoJSON, etc.) that GMT and PyGMT do not support natively.
```
Comment (weiji14) [inline examples/get_started/04_table_inputs.py:97]:
```suggestion
# (e.g., GeoJSON, etc.) that GMT and PyGMT does not support natively.
```

File: examples/get_started/04_table_inputs.py [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+#   is a special syntax in GMT to indicate that the file is a remote file
+#   hosted on the GMT data server.
+
+# %%
+# 2-D array: list, :class:`numpy.ndarray`, and :class:`pandas.DataFrame`
+# ----------------------------------------------------------------------
+#
+# The ``data`` parameter also accepts a 2-D array, e.g.,
+#
+# - A 2-D list (i.e., a list of lists)
```
[the remark this answers is not in the document]
Comment (seisman) [reply examples/get_started/04_table_inputs.py]: Done in 524454973c3c083233237b195337b693a27971bd.

File: examples/get_started/04_table_inputs.py [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+#
+# - A :class:`pathlib.Path` object.
+# - A full URL. PyGMT will download the file to the current directory first.
+# - A file name prefixed with ``@`` (e.g., ``data="@input_data.dat"``), which
+#   is a special syntax in GMT to indicate that the file is a remote file
+#   hosted on the GMT data server.
+
+# %%
+# 2-D array: list, :class:`numpy.ndarray`, and :class:`pandas.DataFrame`
+# ----------------------------------------------------------------------
```
Comment (weiji14) [inline examples/get_started/04_table_inputs.py:65]:
```suggestion
# 2-D array: :class:`list`, :class:`numpy.ndarray`, and :class:`pandas.DataFrame`
# -------------------------------------------------------------------------------
```

File: examples/get_started/04_table_inputs.py [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+
+# %%
+# Conclusion
+# ----------
+#
+# In PyGMT, you have the flexibility to provide data in various table input
+# types, including file names, 2-D arrays (2-D list, :class:`numpy.ndarray`,
+# :class:`pandas.DataFrames`), scalar values or a series of 1-D arrays, and
+# :class:`geopandas.GeoDataFrame`. Choose the input type that best suits your
+# data source and analysis requirements.
```
Comment (weiji14) [inline examples/get_started/04_table_inputs.py:146]:
```suggestion
# types, including file names, 2-D arrays (2-D :class:`list`,
# :class:`numpy.ndarray`, :class:`pandas.DataFrames`), scalar values or a
# series of 1-D arrays, and :class:`geopandas.GeoDataFrame`. Choose the input
# type that best suits your data source and analysis requirements.
```

File: examples/get_started/04_table_inputs.py [hunk 33]
```diff
@@ -0,0 +1,146 @@
+"""
+4. PyGMT I/O: Table inputs
+==========================
```
Comment (weiji14) [inline examples/get_started/04_table_inputs.py:3]: I/O means Input/Output, but this tutorial is only on inputs :slightly_smiling_face: Will the 'Output' part be added as a separate page? Or do we want multiple parts like:

- 4.1 PyGMT I/O: Table inputs
- 4.2 PyGMT I/O: Table outputs
- 4.3 PyGMT I/O: Grid inputs
- 4.4 PyGMT I/O: Grid outputs

Comment (seisman) [reply examples/get_started/04_table_inputs.py:3]: Yes, that's my plan, but I may prefer to the following order:

4.1 PyGMT I/O: Table inputs
4.2 PyGMT I/O: Grid inputs
4.3 PyGMT I/O: Table outputs
4.4 PyGMT I/O: Grid outputs

**Review by weiji14 — APPROVED** [review]
Thanks @seisman! Might need to update this once we get more PyArrow support (#2800), but good to start with the basics first!

**Pull request merged**